### ISSUE TITLE
Improve cycle time

### ISSUE THREAD

**Issue by TiTo-418** [twizmwazin/CardinalPGM#525]
Video proof: http://youtu.be/DzGuRYx3rSw
Version: 18fb2d4f64c3c95180fa2f61ed17de9b8f61c1f2

**Comment by twizmwazin** [comment]
This is primarily due to the drive speed of the server. I'm not sure if there is a safe workaround to pre-copy the world.

**Comment by CaptainElliott** [comment]
I can't think of anything that can be done about this, your server is not powerful enough

**Comment by twizmwazin** [comment]
@CaptainElliott I wouldn't use "powerful," as it has very little to do with the processing speed. The main cause of the cycle time is the disk's IO speed. By moving my testing server from a HDD to an SSD, the cycle time was drastically decreased.

**Comment by SungMatt** [comment]
@twizmwazin I've just did a test on my computer, with a USB disk and the onboard SSD. 

The read/write speed of my USB drive is about 4MB/s / 20MB/s.
The read write speed of my SSD is about 200 MB/s / 450MB/s

Strangely though, the results I'm getting between the two seems to be within reasonable deviation, and within the limits of the terminal time output (It's a bad measure, but the random error of using an external stopwatch is a bit too high)

I've also found out that the second load time is considerably faster than the first.

Here are my results:
|               | Lines of XML | File Size | USB Run 1  | SSD Run 1 | ΔT | USB Run 2 | SSD Run 2 | ΔT |
|---------------|--------------|-----------|------------|-----------|----|-----------|-----------|----|
| Torn Canyon   | 134          | 3.6MB     | 7 seconds  | 8 seconds | +1 | 3 seconds | 3 seconds | ±0 |
| Full Salvo    | 56           | 5.6MB     | 6 seconds  | 4 seconds | -2 | 4 seconds | 3 seconds | -1 |
| Dust          | 102          | 9.5MB     | 4 seconds  | 5 seconds | +1 | 3 seconds | 2 seconds | -1 |
| Warlock       | 81           | 2.8MB     | 0 seconds  | 0 seconds | ±0 | 0 seconds | 1 seconds | +1 |
| BipBetaMC TDM | 90           | 4.4MB     | 10 seconds | 9 seconds | -1 | 4 seconds | 4 seconds | ±0 |

**Comment by twizmwazin** [comment]
@SungMatt Thanks for this data. Markdown is supported in comments as far as I know.

**Comment by twizmwazin** [comment]
@TiTo-418 What is the size of your map's folder? I'm thinking your issue may be partially that the map is not correctly pruned.

**Comment by TiTo-418** [comment]
7,26MB

**Comment by twizmwazin** [comment]
@TiTo-418 Did you prune the world in MCEdit?

**Comment by TiTo-418** [comment]
@twizmwazin Yes, the map is a public dowload of Fractal, fell free to dowload and test

**Comment by TiTi-418** [comment]
Map dowload: https://oc.tc/forums/topics/54075958c2963593d30015a7?page=1#5407750fc29635d91e001b27
My server: https://www.hosthorde.com/plans/ (Unlimited plan)

**Comment by TiTo-418** [comment]
"This is primarily due to the drive speed of the server"

**Comment by twizmwazin** [comment]
Leaving this open as it should be a future goal, optimizations can be made.

**Comment by Pablete1234** [comment]
I'll make a note on this, i ran some tests, on my raspberry, with a modified version of cardinal that sent some debug messages for timing accuarcy.

One of the cycle results (they usually follow the same amounts of time):
Copying the files : 58ms
Creating world : 13 054 ms
Unregistering old modules: 33ms
Creating new modules: 301ms
Handling cycle event: 176ms
Unloading old world: 4 399ms

**Comment by SungMatt** [comment]
@Pablete1234 Did you really try to run Cardinal on your RPi? Wow.

**Comment by twizmwazin** [comment]
@SungMatt I actually think running the tests on a Raspberry Pi is a really good way to measure performance. Because the Pi is much slower than anything intended to run Cardinal, the performance issues will be exaggerated. Then, if Cardinal can be patched to preform well there, it will be extremely quick on an actual production machine.

**Comment by twizmwazin** [comment]
On another note, I think it wouldn't be a bad idea to look into loading an unloading words async. The world loading could be attempted in advance. We could also look into performance improvements for Sportbukkit.

**Comment by Pablete1234** [comment]
Another note:
Cycle time is definetly linked to cpu power, not to ram, RB has 1GB of ram, and i opened the MC server with 512MB, today I tried DitigalOcean VPS's (you can pay them hourly), and the smallest plan (512MB of ram for the whole VPS running fedora) ran super smooth with cardinal

**Comment by twizmwazin** [comment]
If anyone by chance has access to a system where this can be more thoroughly tested, I'd really like to see results. Is there an ARM board <$50 USD that can easily set the CPU frequency? I think that would be a good way to test and keep all other conditions identical.

**Comment by SungMatt** [comment]
@twizmwazin The RPi 2 is able to go from 600MHz to 1GHz. I can do long term testing if you want to induce the effect of bad garbage collection.

**Comment by SungMatt** [comment]
@Pablete1234 What modifications did you make to time the plugin? Is it a branch on your fork?

**Comment by Pablete1234** [comment]
@SungMatt i don't think i have that code anymore, but it was just some messages with times printed to the console

**Comment by Pablete1234** [comment]
@SungMatt i found the code, i rebased it and its on my debug branch, you can take it for testing purposes

**Comment by Pablete1234** [comment]
I ran some tests today, i took Nextgen TE, and made a "pruned" and a "noprune" map out of it, the first one is perfectly pruned, and the second one generated with mcedit 20 chunks of air in each directon.

I tested with 2 cardinal versions, one is regular cardinal (with debug messages for timings) and the other one, i made GameHandler a listener, and onWorldInit event, i setKeepSpawnInMemory(false), so that it doesn't load spawn-chunks.

All the tests were ran in a raspberry pi 2, the MC server was started with 768M of ram:

With spawn chunks loaded (as cardinal is RN):

|        | Pruned | Not pruned |
|:--------:|:---------:|:-----------:|
| Test 1 |  14682 |    7760    |
| Test 2 |  15146 |    7386    |
| Test 3 |  15313 |    7384    |

With spawn chunk loading disabled:

|        | Pruned | Not pruned |
|:--------:|:-------:|:-----------:|
| Test 1 |  5306 |    5685    |
| Test 2 |  5736 |    5705    |
| Test 3 |  5587 |    5269    |


The results were not what i expected at the start, since the non pruned map loads FASTER than the properly pruned map, so it may be better to have non-pruned maps on server that have bad cpu's and lots of HDD space. With the second version stuff changed arround, and both pruned and non pruned maps became alot faster, and kinda equal in timings.

**Comment by twizmwazin** [comment]
I'd be curious to see how this changed if you installed a very slow SD card. Any chance you could provide information about yours?

**Comment by Pablete1234** [comment]
@twizmwazin all i know and i can say is its a "default" mobile micro sd card that came with one of my old mobile devices, still, i don't think SD read time is an issue at all, since map copying is ~40-60ms for the pruned map, and ~100-110ms with the kinda big map

**Comment by SungMatt** [comment]
I do have a really old SD card with a whopping 256 MB of storage, from back in 2006. Can test on that if you want a really slow read/write speed.

**Comment by twizmwazin** [comment]
If you have free time and would like to test, feel free. Otherwise, I'm just going to go off of @Pablete1234's word.
